### ISSUE TITLE
false positive on macro-containing Excel file?

### ISSUE THREAD

**Issue by jacksomervell** [phpMussel/phpMussel#106]
Hi there

Firstly, I love phpMussel - great work.

I have encountered one issue however: I'm looking to ensure Excel files containing Macros are blocked. However, even when enabling VirusTotal with an API key (all I have to do is add the key to the ini file to enable it, right?), they still get through.

Any ideas?

Thanks a lot

Jack

**Comment by DanielRuf** [comment]
@jacksomervell VirusTotal just flags files as malicious which are also flagged as malicious by the used AV scanners. If a scan of the file with virustotal.com does not show any issues, then the file is ok and phpMussel will not block it in most cases.

I am not sure if we have even any own rules in phpMussel to detect Macros and flag / block these files.

You can try to change the suspicion level to make sure that the file is scanned by VirusTotal
https://github.com/Maikuolan/phpMussel/blob/master/vault/phpmussel.ini#L532

When you are logged in to your VirusTotal account you can go to the API requests page and see if there was any request sent to VirusTotal.

**Comment by Maikuolan** [comment]
> Firstly, I love phpMussel - great work.

Thanks! Glad to hear it. :-)

> VirusTotal just flags files as malicious which are also flagged as malicious by the used AV scanners. If a scan of the file with virustotal.com does not show any issues, then the file is ok and phpMussel will not block it in most cases.

This is correct, and possibly may be the issue here. Also, currently (and I hope to improve on this in the future),  the way that phpMussel leverages the VirusTotal API is by checking for existing reports in the Virus Total database (via hashes, etc), rather than by sending any actual complete files to the API (this is because for new files which Virus Total hasn't previously encountered, it will need to actually scan the files against all of its engines, which could take some time, and we haven't yet built any way to handle queuing, scanning in the background, etc; this is on the to-do list, though). What this means is that at this time, if Virus Total has previously encountered a file and flagged it as malicious, phpMussel should also flag it as malicious by way of using the API, but if Virus Total hasn't previously encountered a file, it won't be flagged by the API (yet).

> I am not sure if we have even any own rules in phpMussel to detect Macros and flag / block these files.

We do, _kind_ of.. But to a limited extent.

Currently, if a file is detected as being an "archive", and if the extension of the file matches a filetype known to potentially carry macros, a variable will be marked to indicate this, which will then consequently be used when scanning the contents of the archive to determine (alongside some other simple checks; whether the contents carry an extension known to be used for macros and whether the contents are within a specific filesize range) whether its contents should be checked against the "OLE" signatures of phpMussel.

These "OLE" signatures do contain some signatures intended to block some malicious macros, but they aren't comprehensive, either; They should catch some common, more well-known malicious macros, but aren't very in-depth at the moment.

This particular process should catch a good number of macros, but isn't perfect and probably will require refinement in the future (eg, although most modern formats should be detected as such, and therefore should be scanned, some older formats which don't use the standard PK magic number won't be detected at all, and, items flagged to be checked against the OLE signatures, will generally be macros, but sometimes might just be things like XML data contained in documents, rather than macros, etc). This means that, although we can detect a good number of macros, the process used to do so isn't really reliable enough to be able to definitively say "this file contains a macro and therefore should be blocked" without any risk of potentially also blocking macro-less documents that just contain XML metadata and things like that.

Basically, in summary; We're part-way there, but still have some ways to go yet. :-)

**Comment by Maikuolan** [comment]
<strike>(I'll reply again shortly. Need to head out for an hour or so due to some work issues).</strike>

> (all I have to do is add the key to the ini file to enable it, right?)

Correct (for scanning files using the VirusTotal API).

> Any ideas?

I think what we'll need here is a way to detect whether a file is definitively a macro (as opposed to the "could be" approach currently being used). I'll have a play around with this to see if I can come up with anything.

**Comment by Maikuolan** [comment]
_Improving detection of macros is on the to-do list. Leaving this issue open until we've got something committed for this._

**Comment by Maikuolan** [comment]
Very late getting back to this, but I've just committed some changes tonight that should make it easier to detect macros in files scanned by phpMussel, and a new directive has been added allowing users to block files containing macros. I've indicated it as being "experimental", in that further testing would be required to ensure that it can really catch all of the most common types of macros that exist in documents, spreadsheets and so on, but my local testing thus far looks promising.

Anyhow, now that this is technically a supported thing, I'll mark this issue as "resolved" and close it. Possibly there will be improvements in the near future though based on the results of further testing and on whatever else can be discovered.
